Show backtrace on allocation failures when possible

bjorn3 opened the pull request:
And if an allocation while printing the backtrace fails, don't try to print another backtrace as that will never succeed.

Split out of https://github.com/rust-lang/rust/pull/147725 to allow landing this independently of a decision whether or not to remove `-Zoom=panic`.

bjorn3 committed:
Show backtrace on allocation failures when possible

And if an allocation while printing the backtrace fails, don't try to
print another backtrace as that will never succeed.

RalfJung commented:
Cc @Amanieu @m-ou-se

Mark-Simulacrum commented on library/std/src/alloc.rs:
```diff
@@ -354,28 +359,80 @@ fn default_alloc_error_hook(layout: Layout) {
+    let mut lock = crate::sys::backtrace::lock();
+
+    match crate::panic::get_backtrace_style() {
+        // SAFETY: we took out a lock just a second ago.
```
It's not clear to me what this safety comment is on - I don't see any unsafe code here?

bjorn3 replied:
This was copied from the default panic hook. This comment originates from https://github.com/rust-lang/rust/commit/1c8f9bb84d5956abe6671bf77ccaba64505266a5, which doesn't have unsafe code either. cc @jyn514

bjorn3 replied:
Done

Mark-Simulacrum reviewed (approved):
r=me with safety comment removed or clarified, if we don't want review from Mara or Amanieu. (Or think this needs libs-api signoff since it's sort of new public feature).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148020/4b3ba5844e8831c9b3ee5a5643cdff5da0677426..fefc6d2b59af5aa4d8c926cfd417654c7840c456)*

Amanieu commented:
LGTM from the libs side

@bors r=Mark-Simulacrum

bors commented:
:hourglass: Testing commit 4627bff6e65360c49d422ddc36ad882dcbf0d914 with merge 91c7271516a01828458390930108fd01246c056a...
<!-- homu: {"type":"BuildStarted","head_sha":"4627bff6e65360c49d422ddc36ad882dcbf0d914","merge_sha":"91c7271516a01828458390930108fd01246c056a"} -->

bjorn3 committed:
Fix backtrace/line-tables-only.rs test

The content of the baz function has always been on line 5, not line 15.
Presumably it previously accidentally worked as the call in the foo
function is on line 15, but now it likely tail calls to bar.

bjorn3 commented:
Fixed a test that seems like it happened to work by accident before.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19758026538/job/56613389865) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56613389865) [(plain)](https://github.com/rust-lang/rust/commit/66a677a06c2c30f950f5aedc1c43dba8ebf35161/checks/56613389865/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
warning ../../../../../package.json: License should be a valid SPDX license expression
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
error Error: https://registry.yarnpkg.com/strip-ansi/-/strip-ansi-6.0.1.tgz: Request failed "500 Internal Server Error"
    at ResponseError.ExtendableBuiltin (/node/lib/node_modules/yarn/lib/cli.js:696:66)
    at new ResponseError (/node/lib/node_modules/yarn/lib/cli.js:802:124)
    at Request.<anonymous> (/node/lib/node_modules/yarn/lib/cli.js:66750:16)
    at Request.emit (node:events:518:28)
    at module.exports.Request.onRequestResponse (/node/lib/node_modules/yarn/lib/cli.js:142287:10)
    at ClientRequest.emit (node:events:518:28)
    at HTTPParser.parserOnIncomingClient (node:_http_client:698:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
    at TLSSocket.socketOnData (node:_http_client:540:22)
    at TLSSocket.emit (node:events:518:28)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
yarn install did not exit successfully

thread 'main' (59222) panicked at src/tools/rustdoc-gui-test/src/main.rs:63:10:
unable to install browser-ui-test: Custom { kind: Other, error: "yarn install returned exit code exit status: 1" }
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/std/src/panicking.rs:698:5
   1: core::panicking::panic_fmt
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/panicking.rs:80:14
```

</details>

bjorn3 commented:
@bors retry yarnpkg returned an internal server error

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/19763820567/job/56646153145)
Approved by: Mark-Simulacrum
Pushing 9050733395f7c10560d49f549e022f411fa28cea to main...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"main","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/19763820567/job/56646153145"},"merge_sha":"9050733395f7c10560d49f549e022f411fa28cea"} -->

rust-timer commented:
Finished benchmarking commit (9050733395f7c10560d49f549e022f411fa28cea): [comparison URL](https://perf.rust-lang.org/compare.html?start=d645a4c9c563b80048ce5f32845e754a67f11efa&end=9050733395f7c10560d49f549e022f411fa28cea&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d645a4c9c563b80048ce5f32845e754a67f11efa&end=9050733395f7c10560d49f549e022f411fa28cea&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.6% | [0.5%, 0.8%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d645a4c9c563b80048ce5f32845e754a67f11efa&end=9050733395f7c10560d49f549e022f411fa28cea&stat=max-rss)
<details>
<summary>Results (primary -5.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.1% | [-5.1%, -5.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -5.1% | [-5.1%, -5.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d645a4c9c563b80048ce5f32845e754a67f11efa&end=9050733395f7c10560d49f549e022f411fa28cea&stat=cycles:u)
<details>
<summary>Results (secondary -0.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.2%, 2.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d645a4c9c563b80048ce5f32845e754a67f11efa&end=9050733395f7c10560d49f549e022f411fa28cea&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.1%] | 4     |
| Regressions ❌ <br /> (secondary)  | 0.1% | [0.1%, 0.2%] | 35    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.1%] | 4     |
</details>

**Bootstrap**: 473.697s -> 473.019s (-0.14%)
**Artifact size**: 386.90 MiB -> 386.89 MiB (-0.00%)
